docs💥: revamp extension guide (#1494)
interactions-py/interactions.py: docs/src/Guides/20 Extensions.md

# Extensions

## Introduction

> Your code's getting pretty big and messy being in a single, huh? Wouldn't it be nice if you could organise your commands and listeners into separate files?

Reviewer: ```suggestion
Your code's getting pretty big and messy being in a single file, huh? Wouldn't it be nice if you could organise your commands and listeners into separate files?
```

            self.bot = bot

    def setup(bot):
        # yes, the bot does not need to do any special logic - you just need to pass it into the extension
        MyExtension(bot)
    ```

### Events and Commands

You probably want extensions to do a little bit more than just exist though. Most likely, you want some events and commands
in here. Thankfully, they're relatively simple to do. Expanding on the example a bit, a slash command looks like this:

```python
from interactions import Extension, slash_command, SlashContext

class MyExtension(Extension):
    @slash_command()
    async def test(self, ctx: SlashContext):
        await ctx.send("Hello world!")
```

As you can see, they're almost identical to how you declare slash commands in your main bot file, even using the same decorator.
The only difference is the `self` variable - this is the instance of the extension that the command is being called in, and is
standard for functions inside of classes. Events follow a similar principal.

interactions.py will automatically add all commands and events to the bot when you load the extension (discussed later),
so you don't need to worry about that.

#### Accessing the Bot

> You may have noticed that the `self` variable is being used in the example above. This is because the bot is passed to the extension
> when it's loaded and is stored in the `self.bot` variable. This means you can access the bot from anywhere in your extension,

Reviewer 2: ```suggestion
When an extension is loaded, the library automatically sets the `bot` for you. With this in mind, to access your client, use `self.bot`.

Please do note, if your convention is to use `self.client` this is fine and supported natively as well. 
```

Removed the blatant spoonfeed and added useful info... lets be real, we shouldnt explain basic class usage, our policy is we expect basic python knowledge

```python
class MyExtension(Extension):
    @slash_command()
    async def test(self, ctx: SlashContext):
        await ctx.send(f"Hello, I'm {self.bot.user.mention}!")
```

This also allows you to share data between extensions and the main bot itself. `Client` allows storing data in unused attributes,
so you can do something like this:

```python
from interactions import Client

# main.py
bot = Client(...)
bot.my_data = "Hello world"

# extension.py
class MyExtension(Extension):
    @slash_command()
    async def test(self, ctx: SlashContext):
        await ctx.send(f"My data: {self.bot.my_data}!")
```

### Loading the Extension

Now that you've got your extension, you need to load it.

Let's pretend the extension is in a file called `extension.py`, and it looks like the command example:

```python
from interactions import Extension, slash_command, SlashContext

class MyExtension(Extension):
    @slash_command()
    async def test(self, ctx: SlashContext):
        await ctx.send("Hello world!")
```

Now, let's say you have a file called `main.py` in the same directory that actually has the bot in it:

```python
bot = Client(...)
bot.start("token")
```

> To load the extension, you just need to use `bot.load_extension("filename_in_import_style")` before `bot.start`. So, in this case, it would look like this:

Reviewer 2: ```suggestion
To load the extension, you just need to use `bot.load_extension("filename.in_import_style")` before `bot.start`. So, in this case, it would look like this:
```
Suggest adding a dot, to differentiate between snake_case and import

```python
bot = Client(...)
bot.load_extension("extension")
bot.start("token")
```

And that's it! Your extension is now loaded and ready to go.

#### "Import Style"

> You may have noticed that the filename is passed to `load_extension` without the `.py` extension. This is because interactions.py actually does an
> *import* when loading the extension, so whatever string you give it needs to be a valid Python import path. This means that if you have a file structure like this:

Reviewer 2: ```suggestion
In the example above, the filename is passed to `load_extension` without the `.py` extension. This is because interactions.py actually does an
*import* when loading the extension, so whatever string you give it needs to be a valid Python import path. This means that if you have a file structure like this:
```

```
main.py
exts/
    extension.py
```

You would need to pass `exts.extension` to `load_extension`, as that's the import path to the extension file.

#### Reloading and Unloading Extensions

You can also reload and unload extensions. To do this, you use `bot.reload_extension` and `bot.unload_extension` respectively.

```python
bot.reload_extension("extension")
bot.unload_extension("extension")
```

Reloading and unloading extensions allows you to edit your code without restarting the bot, and to remove extensions you no longer need.
For example, if you organize your extensions so that moderation commands are in one extension, you can reload that extension (and so only moderation-related commands)
as you edit them.

### Initialization

You may have some logic to do when loading a specific extension. For that, you can add the `__init__` method, which takes a `Client` instance, in your extension:

```python
class MyExtension(Extension):
    def __init__(self, bot):
        # do some initialization here
        pass
```

#### Asynchronous Initialization

> You may have noticed `__init__` is synchronous. This may pose problems if you're trying to do something asynchronous in it, so there are various ways of solving it.

Reviewer 2: ```suggestion
As normal, `__init__` is synchronous. This may pose problems if you're trying to do something asynchronous in it, so there are various ways of solving it.
```

        pass
```

!!! warning "Warning about `asyncio.create_task`"
    `asyncio.create_task` only works *if there is an event loop.* For the sake of simplicity we won't discuss what that is too much,
    but the loop is only created when `asyncio.run()` is called (as it is in `bot.start()`). This means that if you call `asyncio.create_task`
    before `bot.start()`, it will not work. If you need to do asynchronous logic before the bot starts, you'll need to load the extension
    in an asynchronous function and use `await bot.astart()` instead of `bot.start()`.

    For example, this format of loading extensions will allow you to use `asyncio.create_task`:

    ```python
    bot = Client(...)

    async def main():
        # event loop made!
        bot.load_extension("extension")
        await bot.astart("token")

    asyncio.run(main())
    ```

### Cleanup

You may have some logic to do while unloading a specific extension. For that, you can override the `drop` method in your extension:

```python
class MyExtension(Extension):
    def drop(self):
        # do some cleanup here
        super().drop()  # important - this part actually does the unloading
```

> You may have noticed that `drop` is synchronous. If you need to do something asynchronous, you can create a task with `asyncio` to do it:

Reviewer 2: ```suggestion
The `drop` method is synchronous. If you need to do something asynchronous, you can create a task with `asyncio` to do it:
```

    bot.load_extension(extension)
```

Note that `glob.glob` returns the *filenames* of all files that match the pattern we provided. To turn it into a module/import name, we need to remove the ".py" suffix and replace the slashes with periods.
On Windows, you may need to replace the slashes with backslashes instead.

???+ note "Note About Loading Extensions From a File"
    While these are two possible ways, they are by no means the *only* ways of finding all extensions in the folder and loading them. Which method is best method depends on your use case and is purely subjective.

### The `setup`/`teardown` Function

You may have noticed that the `Extension` in the extension file is simply just a class, with no way of loading it. interactions.py is smart enough to detect `Extension` subclasses
and use them when loading from a file, but if you want more customization when loading an extension, you'll need to use the `setup` function.

The `setup` function should be *outside* of any `Extension` subclass, and takes in the bot instance, like so:

```python
class MyExtension(Extension):
    ...

def setup(bot):
    # insert logic here
    MyExtension(bot)
```

> Notice how the `Extension` subclass is created inside the `setup` function, and does not need to do any special function to add the extension in beyond being created using the instance.

Reviewer 2: ```suggestion
Here the `Extension` subclass is instantiated inside the `setup` function, and does not need to do any special function to add the extension in beyond being created using the instance.
```
It's not being created, that was done 3 lines above

A similar function can be used for cleanup, called `teardown`. It takes no arguments, and should be outside of any `Extension` subclass, like so:

```python
class MyExtension(Extension):
    ...

def teardown():
    # insert logic here
    pass
```

You usually do not need to worry about unloading the specific extensions themselves, as interactions.py will do that for you.

### Passing Arguments to Extensions

If you would like to pass more than just the bot while initializing an extension, you can pass keyword arguments to the `load_extension` method:

```python
class MyExtension(Extension):
    def __init__(self, bot, some_arg: int = 0):
        ...

bot.load_extension("extension", some_arg=5)
```

If you're using a `setup` function, the argument will be passed to that function instead, so you'll need to pass it to the `Extension` subclass yourself:

```python
class MyExtension(Extension):
    ...

def setup(bot, some_arg: int = 0):
    MyExtension(bot, some_arg)
```

## Extension-Wide Checks

> Sometimes, it is useful to have a check ran before running any command in an extension. Thankfully, all you need to do is use `add_ext_check`:

Reviewer 2: ```suggestion
Sometimes, it is useful to have a check run before running any command in an extension. Thankfully, all you need to do is use `add_ext_check`:
```

        ...
```

### Pre And Post Run Events

Pre- and post-run events are similar to checks. They run before and after a command is invoked, respectively:

```python
from interactions import BaseContext

class MyExtension(Extension):
    def __init__(self, bot: Client):
        self.add_extension_prerun(self.pre_run)
        self.add_extension_postrun(self.post_run)

    async def pre_run(ctx: BaseContext):
        print(f"Command started at: {datetime.datetime.now()}")

    async def post_run(ctx: BaseContext):
        print(f"Command done at: {datetime.datetime.now()}")

    @slash_command(...)
    async def my_command(...):
        # pre and post run will be ran before/after this command
        ...
```

### Extension-Wide Error Handlers

Sometimes, you may want to have a custom error handler for all commands in an extension. You can do this by using `set_extension_error`:

```python
class MyExtension(Extension):
    def __init__(self, bot: Client):
        self.set_extension_error(self.error_handler)

    async def error_handler(self, error: Exception, ctx: BaseContext):
        # handle the error here
        ...
```

??? note "Error Handling Priority"

>     Command error handlers will be ran over extension error handlers, and extension error handlers will be ran over event error handlers.
>     Only one error handler will be ran.

Reviewer 2: ```suggestion
    Only one error handler will run. Similar to css, the most specific handler takes precedence. So in order of priority command error handlers -> extension -> listeners
```
